keyboard/src/lib.rs

```diff
@@ -1,11 +1,433 @@
+                   (K7, false) => '7',          (K7, true) => '&',
+                   (K8, false) => '8',          (K8, true) => '*',
+                   (K9, false) => '9',          (K9, true) => '(',
+           (KBackslash, false) => '\\', (KBackslash, true) => '|',
+            (KBacktick, false) => '`',   (KBacktick, true) => '~',
+               (KComma, false) => ',',      (KComma, true) => '<',
+              (KPeriod, false) => '.',     (KPeriod, true) => '>',
+               (KBackspace, _) => '\x08',
+                   (KEnter, _) => '\n',
+                   (KSpace, _) => ' ',
+                     (KTab, _) => '\t',
+                             _ => return None,
+        };
+
+        Some(character)
+    }
+}
+
+/// Decode a code in the PS/2 scan code set 1 (legacy set).
+///
+/// Difference between set 1 and sets 2 & 3:
+///   http://wiki.osdev.org/%228042%22_PS/2_Controller#Translation
+///
+/// Reference table:
+///   http://www.computer-engineering.org/ps2keyboard/scancodes1.html
+pub fn from_scancode(code: u8) -> Option<(Key, Action)> {
+    use Key::*;
+
+    // Set 1 has the release codes for the basic keys offset by 0x80 from the
+    // press codes for the key
+    let (code, action) = match code {
+        0x00...0x58 => (code,        Action::Pressed),
+        0x81...0xD8 => (code - 0x80, Action::Released),
+        _ => return None,
+    };
+
+    let key = match code {
+        0x1e => KA,
+        0x30 => KB,
+        0x2e => KC,
+        0x20 => KD,
+        0x12 => KE,
+        0x21 => KF,
+        0x22 => KG,
+        0x23 => KH,
+        0x17 => KI,
+        0x24 => KJ,
+        0x25 => KK,
+        0x26 => KL,
+        0x32 => KM,
+        0x31 => KN,
+        0x18 => KO,
+        0x19 => KP,
+        0x10 => KQ,
+        0x13 => KR,
+        0x1f => KS,
+        0x14 => KT,
+        0x16 => KU,
+        0x2f => KV,
+        0x11 => KW,
+        0x2d => KX,
+        0x15 => KY,
+        0x2c => KZ,
+        0x0b => K0,
+        0x02 => K1,
+        0x03 => K2,
+        0x04 => K3,
+        0x05 => K4,
+        0x06 => K5,
+        0x07 => K6,
+        0x08 => K7,
+        0x09 => K8,
+        0x0a => K9,
+        0x29 => KBacktick,
+        0x0c => KDash,
+        0x0d => KEqual,
+        0x0f => KTab,
+        0x2b => KBackslash,
+        0x39 => KSpace,
+        0x1a => KLeftSquareBracket,
+        0x1b => KRightSquareBracket,
+        0x27 => KSemicolon,
+        0x28 => KApostrophe,
+        0x33 => KComma,
+        0x34 => KPeriod,
+        0x35 => KSlash,
+        0x37 => KKeypadAsterisk, // Keypad
+        0x4a => KKeypadMinus,    // Keypad
+        0x4e => KKeypadPlus,     // Keypad
+        0x0e => KBackspace,
+        0x1c => KEnter,
+        0x2A => KLeftShift,
+        0x36 => KRightShift,
+        _    => return None,
+    };
+
+    Some((key, action))
+}
+
+/// Size of the buffer used for storing the state of the keyboard
+const KEY_BUFFER: usize = 256;
```

Comment:
In general, since we don't have a heap yet, this seems like the way to go. But I also wonder if there isn't something better. Hm.